### PR TITLE
[mdns-avahi] handle freed `AvahiWatch/Timeout` entries as processing run loop

### PR DESCRIPTION
**Pull request by abtink**
Avahi provides a mechanism to integrate it to a `select()` style run loop where an implementation can provide an `AvahiWatch` and an `AvahiTimeout` structure along with a set of related APIs (as function pointers given to Avahi client) for it to allocate new watch/timeout entries or update or free them. A watch entry provides a set of events to watch for and expects its given callback to be invoked when the events occur. A timeout provides an abstraction for a timer.

This commit updates `AvahiPoller` (which integrates Avahi to run loop as a `MainloopProcessor`) to protect against situations where entries get freed or updated as Avahi callbacks are invoked from `AvahiPoller::Process()`.

When we invoke the callback for an `AvahiWatch` or `AvahiTimeout` the Avahi module can call any of APIs we provided to it. For example, it can update or free any existing `AvahiWatch/Timeout` entry, which in turn, modifies the `mWatches` or `mTimers` list being tracked by the `AvahiPoller`.

This commit updates the `AvahiPoller` class to correctly handle such situations. Before invoking the callback, we update the entry's state and after returning from callback we restart the iteration over the watch/timer list to find the next entry to report, as the list may have changed during execution of the callback.

### CODE REVIEW

File: src/mdns/mdns_avahi.cpp [hunk 1]
```diff
@@ -364,56 +380,90 @@ void AvahiPoller::Update(MainloopContext &aMainloop)
...
 
-        avahiTimeout->mCallback(avahiTimeout, avahiTimeout->mContext);
+        for (AvahiTimeout *timer : mTimers)
+        {
+            if (timer->mShouldReport)
+            {
+                shouldReport         = true;
+                timer->mShouldReport = false;
+                timer->mCallback(timer, timer->mContext);
+            }
```
Comment (superwhd) [inline src/mdns/mdns_avahi.cpp:465]: Should there be a break?

File: src/mdns/mdns_avahi.cpp [hunk 2]
```diff
@@ -364,56 +380,90 @@ void AvahiPoller::Update(MainloopContext &aMainloop)
...
+        }
```
Comment (superwhd) [inline src/mdns/mdns_avahi.cpp:466]: Probably we can put the iteration of watches and timers into the same `while (shouldReport)`. I'm not sure which version is more readable though.

File: src/mdns/mdns_avahi.cpp (hunk 1 above)
Comment (abtink) [reply src/mdns/mdns_avahi.cpp:465]: Yes. Indeed.  I missed this. Thanks.

Comment (abtink) [reply src/mdns/mdns_avahi.cpp:467]: Fixed in new push. Thanks for noticing this.

File: src/mdns/mdns_avahi.cpp (hunk 2 above)
Comment (abtink) [reply src/mdns/mdns_avahi.cpp:468]: I think better to keep them separate to be more efficient during run time. If we combine and use one loop for both watches and timers, every iteration of the outer `while (shouldReport)` loop in order to get to timers we nee to first go over all watches again in the list (even if all watches are already handled).

**Review by superwhd — APPROVED** [review]
LGTM 👍

**Review by jwhui — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**